File: Lab6-Task3.cpp
#define NOMINMAX
#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>
#include <windows.h>

using namespace std;

const double EPS = 0.001;

// 1. ПРЯМОЙ МЕТОД: Метод Гаусса с выбором главного элемента
vector<double> solveGauss(vector<vector<double>> A, vector<double> B) {
    int n = A.size();
    
    // Прямой ход
    for (int i = 0; i < n; i++) {
        // Поиск максимального элемента в столбце для устойчивости
        int maxRow = i;
        for (int k = i + 1; k < n; k++) {
            if (abs(A[k][i]) > abs(A[maxRow][i])) {
                maxRow = k;
            }
        }
        // Меняем строки местами
        swap(A[i], A[maxRow]);
        swap(B[i], B[maxRow]);

        // Обнуление элементов под главной диагональю
        for (int k = i + 1; k < n; k++) {
            double factor = A[k][i] / A[i][i];
            for (int j = i; j < n; j++) {
                A[k][j] -= factor * A[i][j];
            }
            B[k] -= factor * B[i];
        }
    }

    // Обратный ход
    vector<double> x(n);
    for (int i = n - 1; i >= 0; i--) {
        x[i] = B[i];
        for (int j = i + 1; j < n; j++) {
            x[i] -= A[i][j] * x[j];
        }
        x[i] /= A[i][i];
    }
    return x;
}

// 2. ИТЕРАЦИОННЫЙ МЕТОД: Метод Якоби (Автоматическое вычисление коэффициентов)
vector<double> solveJacobi(const vector<vector<double>>& A, const vector<double>& B, double eps) {
    int n = A.size();
    vector<double> x(n, 0.0);
    vector<double> next_x(n, 0.0);
    int iteration = 0;
    double error;

    cout << setw(3) << "N" << " | " 
         << setw(10) << "X1" << " | " 
         << setw(10) << "X2" << " | " 
         << setw(10) << "X3" << " | " 
         << setw(10) << "X4" << " | " 
         << setw(10) << "En" << "\n";
    cout << "--------------------------------------------------------------\n";

    do {
        // Программа САМА вычисляет x_i через матрицу A и вектор B
        for (int i = 0; i < n; i++) {
            double sum = 0.0;
            for (int j = 0; j < n; j++) {
                if (i != j) {
                    sum += A[i][j] * x[j];
                }
            }
            next_x[i] = (B[i] - sum) / A[i][i];
        }

        // Оценка погрешности
        error = 0.0;
        for (int i = 0; i < n; i++) {
            error = max(error, abs(next_x[i] - x[i]));
        }

        cout << setw(3) << iteration << " | " 
             << setw(10) << fixed << setprecision(5) << next_x[0] << " | " 
             << setw(10) << next_x[1] << " | " 
             << setw(10) << next_x[2] << " | " 
             << setw(10) << next_x[3] << " | " 
             << setw(10) << error << "\n";

        x = next_x;
        iteration++;

        if (iteration > 1000) {
            cout << "Метод не сходится (превышен лимит итераций).\n";
            break;
        }

    } while (error > eps);

    cout << "--------------------------------------------------------------\n";
    cout << "Якоби: Решение найдено за " << iteration << " итераций.\n\n";
    return x;
}

int main() {
    SetConsoleCP(65001);
    SetConsoleOutputCP(65001);

    // Восстановленная матрица A и вектор B из ваших уравнений
    // M = 1.09, P = 0.84
    vector<vector<double>> A = {
        { 1.09, -0.04,  0.21, -1.16},
        { 0.25, -1.23, -0.16, -0.09},
        {-0.21, -0.16,  0.80, -0.13},
        { 0.15, -1.31,  0.06,  0.84}
    };
    
    vector<double> B = {-1.24, 0.84, 2.56, 1.09};

    cout << "=== ПРЯМОЙ МЕТОД (Метод Гаусса) ===\n";
    vector<double> gauss_result = solveGauss(A, B);
    for (size_t i = 0; i < gauss_result.size(); i++) {
        cout << "x" << i + 1 << " = " << fixed << setprecision(6) << gauss_result[i] << "\n";
    }
    cout << "\n";

    cout << "=== ИТЕРАЦИОННЫЙ МЕТОД (Метод Якоби, EPS = " << EPS << ") ===\n";
    vector<double> jacobi_result = solveJacobi(A, B, EPS);
    for (size_t i = 0; i < jacobi_result.size(); i++) {
        cout << "x" << i + 1 << " = " << fixed << setprecision(6) << jacobi_result[i] << "\n";
    }

    return 0;
}

File: Lab6-Task3.go
package main

import (
	"fmt"
	"math"
)

// 1. ПРЯМОЙ МЕТОД: Метод Гаусса
func solveGauss(A [][]float64, B []float64) []float64 {
	n := len(B)
	
	// Копируем срезы, чтобы не испортить исходную матрицу
	aCopy := make([][]float64, n)
	bCopy := make([]float64, n)
	for i := range A {
		aCopy[i] = make([]float64, n)
		copy(aCopy[i], A[i])
		bCopy[i] = B[i]
	}

	// Прямой ход
	for i := 0; i < n; i++ {
		maxRow := i
		for k := i + 1; k < n; k++ {
			if math.Abs(aCopy[k][i]) > math.Abs(aCopy[maxRow][i]) {
				maxRow = k
			}
		}
		aCopy[i], aCopy[maxRow] = aCopy[maxRow], aCopy[i]
		bCopy[i], bCopy[maxRow] = bCopy[maxRow], bCopy[i]

		for k := i + 1; k < n; k++ {
			factor := aCopy[k][i] / aCopy[i][i]
			for j := i; j < n; j++ {
				aCopy[k][j] -= factor * aCopy[i][j]
			}
			bCopy[k] -= factor * bCopy[i]
		}
	}

	// Обратный ход
	x := make([]float64, n)
	for i := n - 1; i >= 0; i-- {
		x[i] = bCopy[i]
		for j := i + 1; j < n; j++ {
			x[i] -= aCopy[i][j] * x[j]
		}
		x[i] /= aCopy[i][i]
	}
	return x
}

// 2. ИТЕРАЦИОННЫЙ МЕТОД: Метод Якоби
func solveJacobi(A [][]float64, B []float64, eps float64) []float64 {
	n := len(B)
	x := make([]float64, n)
	nextX := make([]float64, n)
	iteration := 0
	maxIter := 1000
	var err float64

	fmt.Printf("%-3s | %-10s | %-10s | %-10s | %-10s | %-10s\n", "N", "X1", "X2", "X3", "X4", "En")
	fmt.Println("----------------------------------------------------------------")

	for {
		for i := 0; i < n; i++ {
			sum := 0.0
			for j := 0; j < n; j++ {
				if i != j {
					sum += A[i][j] * x[j]
				}
			}
			nextX[i] = (B[i] - sum) / A[i][i]
		}

		err = 0.0
		for i := 0; i < n; i++ {
			err = math.Max(err, math.Abs(nextX[i]-x[i]))
		}

		fmt.Printf("%-3d | %10.5f | %10.5f | %10.5f | %10.5f | %10.5f\n", iteration, nextX[0], nextX[1], nextX[2], nextX[3], err)

		for i := 0; i < n; i++ {
			x[i] = nextX[i]
		}
		iteration++

		if iteration > maxIter {
			fmt.Println("Метод не сходится (превышен лимит итераций).")
			break
		}
		if err <= eps {
			break
		}
	}

	fmt.Println("----------------------------------------------------------------")
	fmt.Printf("Якоби: Решение найдено за %d итераций.\n\n", iteration)
	return x
}

func main() {
	// Матрица A и вектор B
	A := [][]float64{
		{1.09, -0.04, 0.21, -1.16},
		{0.25, -1.23, -0.16, -0.09},
		{-0.21, -0.16, 0.80, -0.13},
		{0.15, -1.31, 0.06, 0.84},
	}
	B := []float64{-1.24, 0.84, 2.56, 1.09}
	eps := 0.001

	fmt.Println("=== ПРЯМОЙ МЕТОД (Метод Гаусса) ===")
	gaussResult := solveGauss(A, B)
	for i, val := range gaussResult {
		fmt.Printf("x%d = %.6f\n", i+1, val)
	}
	fmt.Println()

	fmt.Printf("=== ИТЕРАЦИОННЫЙ МЕТОД (Метод Якоби, EPS = %.3f) ===\n", eps)
	jacobiResult := solveJacobi(A, B, eps)
	for i, val := range jacobiResult {
		fmt.Printf("x%d = %.6f\n", i+1, val)
	}
}
